File: libs/httpcl/src/oauth1-signature.cpp
#include "httpcl/oauth1-signature.hpp"

#include <openssl/evp.h>
#include <openssl/hmac.h>
#include <openssl/buffer.h>
#include <httplib.h>
#include <random>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <chrono>
#include <stdexcept>

namespace httpcl
{
namespace oauth1
{

std::string generateNonce(int length)
{
    if (length < 8 || length > 64) {
        throw std::runtime_error("Nonce length must be between 8 and 64");
    }

    static const char alphanum[] =
        "0123456789"
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "abcdefghijklmnopqrstuvwxyz";

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, sizeof(alphanum) - 2);

    std::string nonce;
    nonce.reserve(length);
    for (int i = 0; i < length; ++i) {
        nonce += alphanum[dis(gen)];
    }
    return nonce;
}

std::string generateTimestamp()
{
    auto now = std::chrono::system_clock::now();
    auto epoch = now.time_since_epoch();
    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(epoch);
    return std::to_string(seconds.count());
}

/**
 * RFC 3986 percent encoding.
 * OAuth 1.0 requires unreserved characters (A-Z, a-z, 0-9, -, ., _, ~) to NOT be encoded.
 * All other characters MUST be encoded.
 */
static std::string percentEncode(const std::string& input)
{
    std::ostringstream encoded;
    encoded.fill('0');
    encoded << std::hex << std::uppercase;

    for (unsigned char c : input) {
        // RFC 3986 unreserved characters
        if (std::isalnum(c) || c == '-' || c == '.' || c == '_' || c == '~') {
            encoded << c;
        } else {
            // Percent-encode everything else
            encoded << '%' << std::setw(2) << static_cast<int>(c);
        }
    }

    return encoded.str();
}

/**
 * Base64 encode using OpenSSL.
 */
static std::string base64Encode(const unsigned char* input, size_t length)
{
    BIO* bio = BIO_new(BIO_s_mem());
    BIO* b64 = BIO_new(BIO_f_base64());
    BIO_set_flags(b64, BIO_FLAGS_BASE64_NO_NL);
    bio = BIO_push(b64, bio);

    BIO_write(bio, input, static_cast<int>(length));
    BIO_flush(bio);

    BUF_MEM* bufferPtr;
    BIO_get_mem_ptr(bio, &bufferPtr);

    std::string result(bufferPtr->data, bufferPtr->length);
    BIO_free_all(bio);

    return result;
}

/**
 * Compute HMAC-SHA256.
 */
static std::string hmacSha256(const std::string& key, const std::string& data)
{
    unsigned char digest[EVP_MAX_MD_SIZE];
    unsigned int digestLen = 0;

    HMAC(EVP_sha256(),
         key.data(), static_cast<int>(key.size()),
         reinterpret_cast<const unsigned char*>(data.data()), data.size(),
         digest, &digestLen);

    return std::string(reinterpret_cast<char*>(digest), digestLen);
}

/**
 * Build signature base string according to RFC 5849 Section 3.4.1.
 */
static std::string buildSignatureBaseString(
    const std::string& httpMethod,
    const std::string& url,
    const std::map<std::string, std::string>& params)
{
    // 1. Normalize parameters: sort and concatenate
    std::vector<std::string> encodedPairs;
    for (const auto& [key, value] : params) {
        encodedPairs.push_back(percentEncode(key) + "=" + percentEncode(value));
    }
    std::sort(encodedPairs.begin(), encodedPairs.end());

    std::string paramString;
    for (size_t i = 0; i < encodedPairs.size(); ++i) {
        if (i > 0) paramString += "&";
        paramString += encodedPairs[i];
    }

    // 2. Build signature base string: METHOD&URL&PARAMS
    std::string baseString;
    baseString += httpMethod;
    baseString += "&";
    baseString += percentEncode(url);
    baseString += "&";
    baseString += percentEncode(paramString);

    return baseString;
}

std::string computeSignature(
    const std::string& httpMethod,
    const std::string& url,
    const std::map<std::string, std::string>& params,
    const std::string& consumerSecret,
    const std::string& tokenSecret)
{
    // Build signature base string
    std::string baseString = buildSignatureBaseString(httpMethod, url, params);

    // Build signing key: consumer_secret&token_secret
    std::string signingKey = percentEncode(consumerSecret) + "&" + percentEncode(tokenSecret);

    // Compute HMAC-SHA256
    std::string hmac = hmacSha256(signingKey, baseString);

    // Base64 encode
    return base64Encode(reinterpret_cast<const unsigned char*>(hmac.data()), hmac.size());
}

std::string buildAuthorizationHeader(
    const std::string& httpMethod,
    const std::string& url,
    const std::string& consumerKey,
    const std::string& consumerSecret,
    const std::map<std::string, std::string>& bodyParams,
    int nonceLength)
{
    // Generate OAuth parameters
    std::string timestamp = generateTimestamp();
    std::string nonce = generateNonce(nonceLength);

    // Build combined parameters for signature (OAuth + body params)
    std::map<std::string, std::string> allParams;
    allParams["oauth_consumer_key"] = consumerKey;
    allParams["oauth_signature_method"] = "HMAC-SHA256";
    allParams["oauth_timestamp"] = timestamp;
    allParams["oauth_nonce"] = nonce;
    allParams["oauth_version"] = "1.0";

    // Add body parameters to signature
    for (const auto& [key, value] : bodyParams) {
        allParams[key] = value;
    }

    // Compute signature
    std::string signature = computeSignature(httpMethod, url, allParams, consumerSecret);

    // Build Authorization header (OAuth params only, not body params)
    std::ostringstream header;
    header << "OAuth ";
    header << "oauth_consumer_key=\"" << percentEncode(consumerKey) << "\", ";
    header << "oauth_signature_method=\"HMAC-SHA256\", ";
    header << "oauth_timestamp=\"" << timestamp << "\", ";
    header << "oauth_nonce=\"" << percentEncode(nonce) << "\", ";
    header << "oauth_version=\"1.0\", ";
    header << "oauth_signature=\"" << percentEncode(signature) << "\"";

    return header.str();
}

}  // namespace oauth1
}  // namespace httpcl

File: libs/jzswag-desktop/src/main/java/com/ndsev/zswag/desktop/OAuth1Signature.java
package com.ndsev.zswag.desktop;

import org.jetbrains.annotations.NotNull;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.InvalidKeyException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * OAuth 1.0 (RFC 5849) signature utilities — HMAC-SHA256 only. Java port of
 * C++ {@code httpcl::oauth1::*}. Used for the
 * {@code rfc5849-oauth1-signature} variant of OAuth2 token-endpoint
 * authentication.
 */
public final class OAuth1Signature {
    private static final SecureRandom RANDOM = new SecureRandom();
    private static final char[] ALPHANUM =
            ("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz").toCharArray();

    private OAuth1Signature() {}

    /** Cryptographically secure alphanumeric nonce of the given length (8..64). */
    @NotNull
    public static String generateNonce(int length) {
        if (length < 8 || length > 64) {
            throw new IllegalArgumentException("Nonce length must be between 8 and 64");
        }
        StringBuilder sb = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            sb.append(ALPHANUM[RANDOM.nextInt(ALPHANUM.length)]);
        }
        return sb.toString();
    }

    /** Seconds-since-epoch as decimal string. */
    @NotNull
    public static String generateTimestamp() {
        return Long.toString(System.currentTimeMillis() / 1000L);
    }

    /**
     * RFC 3986 percent-encoding: keep unreserved characters (A-Z, a-z, 0-9, -, ., _, ~);
     * percent-encode everything else as upper-case hex.
     */
    @NotNull
    public static String percentEncode(@NotNull String input) {
        byte[] bytes = input.getBytes(StandardCharsets.UTF_8);
        StringBuilder sb = new StringBuilder(bytes.length * 3);
        for (byte b : bytes) {
            int u = b & 0xFF;
            if ((u >= 'A' && u <= 'Z')
                    || (u >= 'a' && u <= 'z')
                    || (u >= '0' && u <= '9')
                    || u == '-' || u == '.' || u == '_' || u == '~') {
                sb.append((char) u);
            } else {
                sb.append('%');
                sb.append(Character.toUpperCase(Character.forDigit((u >> 4) & 0xF, 16)));
                sb.append(Character.toUpperCase(Character.forDigit(u & 0xF, 16)));
            }
        }
        return sb.toString();
    }

    /**
     * Builds the signature base string per RFC 5849 Section 3.4.1:
     * {@code METHOD&percent(URL)&percent(sorted-percent-encoded-params)}.
     */
    @NotNull
    static String buildSignatureBaseString(@NotNull String httpMethod, @NotNull String url,
                                           @NotNull Map<String, String> params) {
        List<String> encodedPairs = new ArrayList<>(params.size());
        for (Map.Entry<String, String> e : params.entrySet()) {
            encodedPairs.add(percentEncode(e.getKey()) + "=" + percentEncode(e.getValue()));
        }
        Collections.sort(encodedPairs);
        StringBuilder paramString = new StringBuilder();
        for (int i = 0; i < encodedPairs.size(); i++) {
            if (i > 0) paramString.append('&');
            paramString.append(encodedPairs.get(i));
        }
        return httpMethod.toUpperCase(Locale.ROOT) + "&" + percentEncode(url) + "&" + percentEncode(paramString.toString());
    }

    /**
     * Computes HMAC-SHA256 signature for a token request.
     * Signing key is {@code percent(consumer_secret)&percent(token_secret)}; for the
     * client-credentials flow {@code token_secret} is empty.
     */
    @NotNull
    public static String computeSignature(@NotNull String httpMethod, @NotNull String url,
                                          @NotNull Map<String, String> params,
                                          @NotNull String consumerSecret, @NotNull String tokenSecret) {
        String base = buildSignatureBaseString(httpMethod, url, params);
        String key = percentEncode(consumerSecret) + "&" + percentEncode(tokenSecret);
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] hmac = mac.doFinal(base.getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(hmac);
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new IllegalStateException("HMAC-SHA256 unavailable: " + e.getMessage(), e);
        }
    }

    /**
     * Builds the {@code Authorization: OAuth ...} header for a signed token
     * request, including all five {@code oauth_*} parameters and the computed
     * signature. Body parameters are included in signature computation but are
     * NOT echoed in the header.
     */
    @NotNull
    public static String buildAuthorizationHeader(
            @NotNull String httpMethod, @NotNull String url,
            @NotNull String consumerKey, @NotNull String consumerSecret,
            @NotNull Map<String, String> bodyParams, int nonceLength) {
        String timestamp = generateTimestamp();
        String nonce = generateNonce(nonceLength);

        Map<String, String> allParams = new LinkedHashMap<>();
        allParams.put("oauth_consumer_key", consumerKey);
        allParams.put("oauth_signature_method", "HMAC-SHA256");
        allParams.put("oauth_timestamp", timestamp);
        allParams.put("oauth_nonce", nonce);
        allParams.put("oauth_version", "1.0");
        allParams.putAll(bodyParams);

        String signature = computeSignature(httpMethod, url, allParams, consumerSecret, "");

        StringBuilder h = new StringBuilder("OAuth ");
        h.append("oauth_consumer_key=\"").append(percentEncode(consumerKey)).append("\", ");
        h.append("oauth_signature_method=\"HMAC-SHA256\", ");
        h.append("oauth_timestamp=\"").append(timestamp).append("\", ");
        h.append("oauth_nonce=\"").append(percentEncode(nonce)).append("\", ");
        h.append("oauth_version=\"1.0\", ");
        h.append("oauth_signature=\"").append(percentEncode(signature)).append("\"");
        return h.toString();
    }
}
